### PR TITLE
Users full name and orcid is shown and filterable during the addition of collaborators to a project

### PR DESCRIPTION
**Pull request by Steffengreiner**
**What was changed**

1.) The full name and orcid information is also shown within the project access grid and the columns are now resizable and moveable

Screenshot 2024-07-26 at 15 50 56

2.) UserInfo now contains the orcId if provided during registration
3.) Adding an user to a project now shows the orcId and full name of the user during the add collaborator screen. (See screenshot)
Screenshot 2024-07-26 at 15 51 05

4.) Users are filterable via their orcid and full name in the add users to project dialog (see recording)
https://github.com/user-attachments/assets/71b6764f-22c2-40fd-b0fc-07bc62e85b69

### CODE REVIEW

File: identity-api/src/main/java/life/qbic/identity/api/UserInfo.java [hunk 1]
```diff
@@ -10,6 +11,6 @@
  * @since 1.0.0
  */
 public record UserInfo(String id, String fullName, String emailAddress, String platformUserName,
-                       boolean isActive) implements Serializable {
+                       boolean isActive, Optional<String> oidcId) implements Serializable {
```
Comment (KochTobi) [inline identity-api/src/main/java/life/qbic/identity/api/UserInfo.java:14]: You should not use Optional in a record. Optional is not serializable and may break the record. Use `null` in a record instead.

File: identity-infrastructure/src/main/java/life/qbic/identity/infrastructure/UserJpaRepository.java [hunk 2]
```diff
@@ -60,13 +61,62 @@ public Optional<User> findUserByUserName(String userName) {
...
+
+    public static Specification<User> isFullName(String filter) {
+      return (root, query, builder) ->
+          builder.like(root.get("fullName"), "%" + filter + "%");
+    }
+
+    public static Specification<User> isOrcid(String filter) {
+      return (root, query, builder) ->
+          builder.like(root.get("oidcId"), "%" + filter + "%");
+    }
```
Comment (KochTobi) [inline identity-infrastructure/src/main/java/life/qbic/identity/infrastructure/UserJpaRepository.java:115]: careful. OpenId Connect (oidc) is not equal to ORCiD!
You might need to filter for the issuer. For example, "tester von und zu QBiC" does not have an ORCiD but uses only their sandbox instance.
You could introduce a property in the applicaiton properties specifying the issuer Url(s) to consider here.

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/project/access/AddCollaboratorToProjectDialog.java [hunk 3]
```diff
@@ -167,6 +174,60 @@ public Registration addConfirmListener(ComponentEventListener<ConfirmEvent> list
...
+      addInfoItem("Username", new Span(userName));
+      addInfoItem("Full Name", new Span(fullName));
+      add(userAvatar, userInfoContent);
+    }
+
+    private void setOrcid(String orcid) {
+      if (orcid.isEmpty()) {
+        return;
+      }
+      String orcidURL = "https://orcid.org/" + orcid;
```
Comment (KochTobi) [inline user-interface/src/main/java/life/qbic/datamanager/views/projects/project/access/AddCollaboratorToProjectDialog.java:200]: consider using the issuer url here

**Review by KochTobi — CHANGES_REQUESTED** [review]

File: identity-api/src/main/java/life/qbic/identity/api/UserInfo.java (hunk 1 above)
Comment (Steffengreiner) [reply identity-api/src/main/java/life/qbic/identity/api/UserInfo.java]: Good point, I adjusted the record to contain a string and check for null values explictly

File: identity-infrastructure/src/main/java/life/qbic/identity/infrastructure/UserJpaRepository.java (hunk 2 above)
Comment (Steffengreiner) [reply identity-infrastructure/src/main/java/life/qbic/identity/infrastructure/UserJpaRepository.java]: True, I renamed the specification and added a comment that it should be extended if multiple oidc are employed for a user. Currently we only want to check if a provided string contains any oidc value

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/project/access/AddCollaboratorToProjectDialog.java (hunk 3 above)
Comment (Steffengreiner) [reply user-interface/src/main/java/life/qbic/datamanager/views/projects/project/access/AddCollaboratorToProjectDialog.java]: It's completely overhauled to account for the possibility of multiple oidcs and reusability

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/project/access/ProjectAccessComponent.java [hunk 4]
```diff
@@ -89,91 +92,137 @@ protected ProjectAccessComponent(
...
 
-  private Grid<ProjectAccessService.ProjectCollaborator> projectCollaboratorGrid() {
+  private void showControls(boolean isVisible) {
+    buttonBar.setVisible(isVisible);
+  }
 
-    Grid<ProjectAccessService.ProjectCollaborator> grid = new Grid<>();
-    Editor<ProjectCollaborator> editor = grid.getEditor();
-    Binder<ProjectCollaborator> binder = new Binder<>(ProjectCollaborator.class);
+  private void initProjectUserGrid() {
```
Comment (KochTobi) [inline user-interface/src/main/java/life/qbic/datamanager/views/projects/project/access/ProjectAccessComponent.java:135]: This grid has a lot of columns.
Can you use the same Component to render a user in the first column only?
I feel like a lot of space is wasted with information about the users and the main focus of this page should be access management (which disappears to the right hand side)
Screenshot 2024-07-22 at 10 35 36

**Review by KochTobi — CHANGES_REQUESTED** [review]
Overall looks good.
Some minor points we should discuss

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/project/access/ProjectAccessComponent.java [hunk 5]
```diff
@@ -89,91 +92,137 @@ protected ProjectAccessComponent(
...
+    addCollaboratorButton.addClickListener(event -> openAddCollaboratorDialog());
+    buttonBar.add(addCollaboratorButton);
     header.add(titleField, buttonBar);
     add(header);
     Span userProjectAccessDescription = new Span("Users with access to this project");
-
-    projectCollaborators = projectCollaboratorGrid();
-    add(userProjectAccessDescription, projectCollaborators);
+    projectUserGrid = new Grid<>(ProjectUser.class);
+    initProjectUserGrid();
```
[the remark this answers is not in the document]
Comment (Steffengreiner) [reply user-interface/src/main/java/life/qbic/datamanager/views/projects/project/access/ProjectAccessComponent.java:103]: Good point i changed the method to create and return the grid

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/project/access/ProjectAccessComponent.java (hunk 4 above)
Comment (Steffengreiner) [reply user-interface/src/main/java/life/qbic/datamanager/views/projects/project/access/ProjectAccessComponent.java:135]: Good point i adapted the frontend component accordingly

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/project/access/ProjectAccessComponent.java [hunk 6]
```diff
@@ -89,91 +92,137 @@ protected ProjectAccessComponent(
...
-    buttonBar.removeAll();
+  private boolean isCurrentUser(ProjectUser projectUser) {
+    var userId = this.authenticationToUserIdTranslator.translateToUserId(
+        SecurityContextHolder.getContext().getAuthentication()).orElseThrow();
+    return Objects.equals(projectUser.userId(), userId);
   }
 
-  private Grid<ProjectAccessService.ProjectCollaborator> projectCollaboratorGrid() {
+  private void showControls(boolean isVisible) {
+    buttonBar.setVisible(isVisible);
```
[the remark this answers is not in the document]
Comment (Steffengreiner) [reply user-interface/src/main/java/life/qbic/datamanager/views/projects/project/access/ProjectAccessComponent.java:132]: I rewrote how the visibility is set as suggested by adding or removing the component dependent on the user access

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/project/access/AddCollaboratorToProjectDialog.java [hunk 7]
```diff
@@ -167,6 +176,52 @@ public Registration addConfirmListener(ComponentEventListener<ConfirmEvent> list
...
+    }
+
+    private void setOidc(String oidcIssuer, String oidc) {
+      if (oidcIssuer.isEmpty() || oidc.isEmpty()) {
+        return;
+      }
+      var oidcType = Arrays.stream(OidcType.values())
+          .filter(ot -> ot.getIssuer().equals(oidcIssuer))
+          .findFirst().orElseThrow(
+              () -> new IllegalArgumentException("Unknown oidc Issuer %s".formatted(oidcIssuer)));
```
[the remark this answers is not in the document]
Comment (Steffengreiner) [reply user-interface/src/main/java/life/qbic/datamanager/views/projects/project/access/AddCollaboratorToProjectDialog.java:204]: Good idea! I changed it accordingly

**Pull request merged**